### ISSUE TITLE
Proposal to add `buildInstrinctions` in schema.org

### ISSUE THREAD

**Issue by moranegg** [codemeta/codemeta#245]
During the last few months we have reviewed the additional 10 properties that aren't in schema.org and are only in CodeMeta, `buildInstructions` is one of these properties.

The definition used on CodeMeta is:
 > link to installation instructions/documentation

Documentation is a main part in software development and is crucial in the understanding of the software and its functionalities. There are different types of documentation:
- user manual
- build instructions
- tutorials

Therefore, when describing a software project or artifact with metadata it is useful to have access to its documentation and its build instructions.

--------------------
Note:

In the crosswalk table at the moment this property isn't used and it might be more appropriate to rename `buildInstructions` to `documentationUrl` before submitting the issue to schema.org.

Another option is to have both metadata properties (which are different).

**Comment by moranegg** [comment]
Today, during the CodeMeta task force call which is part of the FORCE11 Software Citation Implementation working group we have discussed this issue, including @mfenner, @tmorrell, Shelly Stall, Qian Zhang and Natalie Meyers.

We agreed that *documentation* is missing from the current CodeMeta vocabulary and it is important to have this term which is a general term that can include user manuals, build instructions and tutorials.

I have found the *[documentation](https://schema.org/documentation)* pending property on schema.org which is proposed with the *WebAPI* pending property which will be under service and not under software or creativeWork.

Here is the discussion on schema.org about WebAPI https://github.com/schemaorg/schemaorg/issues/1423

For now, I will open a PR for the v3 milestone to add the *documentation* property under CodeMeta terms without deleting the *buildInstructions* property.

**Comment by arfon** [comment]
> Another option is to have both metadata properties (which are different).

My vote would be to try and standardize on `documentation` and drop `buildInstructions` in the long run for two reasons:

1. It limits the size of the vocabulary.
2. IMO `buildInstructions` is just one example of a possible many subsets of a `documentation` property. So if we want to go in that direction, we should probably think about others.

**Comment by dgarijo** [comment]
I would like to keep separate `buildInstructions` from `documentation`. Reasons:
- Documentations may be huge (some software components have hundreds of pages), but build instructions are a concrete part of the documentation. Sometimes buildInstructions are not even a part of the documentation.
- I can use the `buildInstructions` property to point to Dockerfiles, which are machine-readable build instructions. Thanks to this, I can build tools directly from software repos.

@arfon why is a limited vocabulary a plus? I think that it's better to have a comprehensive vocabulary rather than incomplete one. E.g., schema.org is quite big and it continues growing.

@moranegg note also that there is a schema.org property called `softwareHelp`, which could be seen as a tutorial or documentation too. Domain: `https://schema.org/SoftwareApplication`

**Comment by arfon** [comment]
> @arfon why is a limited vocabulary a plus? I think that it's better to have a comprehensive vocabulary rather than incomplete one. E.g., schema.org is quite big and it continues growing.

I agree in theory, but as I currently read, we're proposing only two types of docs: `documentation` and `buildInstructions`. IMO this is an incomplete vocabulary to describe the various types of documentation (user guides, tutorials, etc.) so in the absence of doing this 'right' then I propose we limit ourselves to only `documentation`.

**Comment by dgarijo** [comment]
@arfon we can start with these two types (which I think are sufficiently different, as one is how to build the program, and the other may be seen more as a usability thing) and then expand from there. I will open an issue.

**Comment by arfon** [comment]
> @arfon we can start with these two types (which I think are sufficiently different, as one is how to build the program, and the other may be seen more as a usability thing) and then expand from there. I will open an issue.

I won't fight you on this – I gave my opinion but I think you have a legitimate case for wanting to keep the `buildInstructions` 😸.

**Comment by mfenner** [comment]
Force11 Codemeta Task Force suggests using `documentation`, and not go into subcategories such as `buildInstructions`. But `buildInstructions` should be referred to in the description.

**Comment by dgarijo** [comment]
@mfenner wouldn't that render the property NON machine readable? If we conflate everything under description, then it is not useful anymore. For example, a Dockerfile are `buildInstructions` to automatically run a software component...

**Comment by mfenner** [comment]
We (the Force11 Codemeta Task Force) felt that there are bigger obstacles to making the property machine-readable. Basically that we don't have any community norms. So a lot more work needed, and hopefully something that can be added in a future version of codemeta.

**Comment by tmorrell** [comment]
PR implementing change is #260

**Comment by dgarijo** [comment]
@mfenner @tmorrell, I find this PR problematic as per my comments in https://github.com/codemeta/codemeta/issues/247

@mfenner, I hear your comments about a lack of common practice. However, there are [over a million repositories with Docker files](https://www.docker.com/blog/intro-guide-to-dockerfile-best-practices/), which are build instructions and may be a valid point for starting using this property. This is a current common practice in GitHub and other repositories.

I find this rationale inconsistent with e.g., development_status, for which the group suggests https://www.repostatus.org/. As reported in their page, it has been used by little over 1K projects, which is a few order of magnitude lower than e.g. Docker files. I am not opposing to using repostatus, but I do not understand why it is enough to be considered common practice for status, and my example for Docker is not.

**Comment by moranegg** [comment]
Thank you for the feedback! I'm very happy to have a different comments and
you are right that the property itself is an important piece in the
metadata.

I'm wondering what's the property used in WikiData?

I believe that the idea suggested here is to type the documentation
property with a qualifier or another relationship(like with Role), because
documentation in itself can be many things, yet it is still a piece of
documentation:
- build instructions
- Readme
- link to a website
- link to a repository holding the docs, a documentation module
- manual

It is not about reducing the vocabulary, but improving it semantically;-)

Concerning the 'deprecated' tags, these changes will only be in v3 of the
vocabulary, and files that were or are still created with the v2 namepace
will not be affected.

I'll try test the documentation type proposition with a real example with
the documentation typing when I have time to do so.

>
>

**Comment by dgarijo** [comment]
@moranegg, 
I checked Wikidata, a couple of software repos with Docker files
like https://www.wikidata.org/wiki/Q47509047 and there is no property for Docker instructions/build files

Rather than adding qualifiers, why not use subproperties? E.g.,  build instructions, readme being a subproperty of documentation.
Some reasons:
- Using qualifiers/reification is a more complicated solution both when representing in RDF and JSON-LD.
- Build instructions is not something like a role, which may change. It's permanent. 
- All the features of documentation (domain, range) are inherited by readme/build instructions.

As for having a new version v3, I understand that it has its own entity and URL. however, as someone who has been managing ontologies and vocabularies for a long time, having incompatible versions may bring in problems:
- The namespace for codemeta terms (as in https://github.com/codemeta/codemeta/blob/master/codemeta.jsonld) is https://codemeta.github.io/terms/. This does not reflect the version, and hence data annotated with one version will be indistinguishable. I raised a similar issue here: https://github.com/codemeta/codemeta/issues/216.  
- If the version is added in the context but resolves to https://codemeta.github.io/terms/, then the version schema will be lost e.g., when translating into other RDF serializations, and hence we will have the same problem. In schema.org there is a property to annotate the version of the schema being used to annotate the data to address some of these issues.
- Codemeta aims to be a standard, and therefore it should be backwards compatible. Otherwise, integrating different versions of codemeta will be a nightmare.

**Comment by tmorrell** [comment]
I've closed my PR so there can be more discussion on how to implement this.

**Comment by cboettig** [comment]
Thanks @tmorrell.  FWIW I agree that this is a type of documentation.  I believe each of these types of documentation could be nested under  `documentation` as a URI.  If further distinction is necessary, I believe that would already be provided by describing each documentation URL as a separate  `CreativeWork` (hurrah for linked data -- we have a link to a thing (`documentation`) and we have a pretty rich vocabulary to describe that thing (`CreativeWork`), so we don't need to introduce new terms).


Still, I agree that the backwards compatibility is a concern and worth discussing separately how to namespace and deprecate terms.  (e.g. clearly schema.org does not use a version-stable namespace and also deprecates terms at some point, but I am not clear on the policy or mechanics they use either).

**Comment by dgarijo** [comment]
While I agree that returning a list may be a solution, I think separating resources as a CreativeWork is not enough. Extensions of CreativeWork would be necessary, because then the main point, which is to try to standardize how people refer to build instructions, would not be possible. It would be left to how users name those resources...

In short, when using `buildInstructions` to point to the Dockerfile of a repo, I would really like to distinguish that value from `documentation`.

**Comment by cboettig** [comment]
@dgarijo I don't believe the property `buildInstructions` has enough semantic precision to do what you want it to do.  The term means different things in different contexts and to different developers -- e.g. many products might point to a README or an INSTALL file in a repo in order to have something to put in the field.  Can you identify an operation that machine could perform by knowing only that a given URL was tagged "buildInstructions" and not "documentation"?  

I don't see how using the term "buildInstructions" in place of "documentation" gets us any closer to allowing a machine to address  the implied use case here of distinguishing a project's "Dockerfile" from the project's "readme".  We would have to create a field called "Dockerfile" for that.  I suspect any machine seeking to perform this task would be better off looking at the filename of the target object, `Dockerfile*` or `README*` rather than relying on a dedicated predicate.  

Certainly we could imagine one day having a well-defined ontology of Documentation types extending CreativeWork.  But for now, using CreativeWork to provide further description about documentation (such as filename, type, author, license, etc) is probably the best way to support use cases which require more nuanced understanding of the different documentation sources.

**Comment by dgarijo** [comment]
I don't think it's "in place of", it's "in addition to". It helps distinguish the purpose of the documentation when aggregating information and presenting it to users. May I ask why was this property added to codemeta in the first place? There must have been concrete use cases for supporting it besides the ones I listed in this thread, right?

If the semantics is loose, why can't we define it better?

The reason why I don't want to use something like "Dockerfile*" it's because maybe I want to point to other type of build files such as docker-compose, k8 config files, etc. Or maybe in a few months it's called something different.

**Comment by cboettig** [comment]
>  May I ask why was this property added to codemeta in the first place? There must have been concrete use cases for supporting it besides the ones I listed in this thread, right?

Recall that codemeta was initially proposed  only as a crosswalk between existing vocabularies, not a new vocabulary. Most of those existing vocabularies operated in a more narrowly defined space, e.g. "R packages", in which terms could reliably be given a more precise meaning that does not always easily generalize to other contexts.  Initially there was no general term "documentation", but a variety of more specific terms that we all agree can indeed be generalized as "documentation".  Obviously we don't want a new field for every possible config file (Dockerfile, compose, etc), and surely we can agree that not doing so does not imply that "our semantics are too loose" and should "define it better".  It's merely a question of finding the right level of generality to be robust and functional.

Generalization can be very hard and something we learn as a community over time.  Continuing to only add more specific terms does not seem ideal: should we add a new property for "tutorials" as well?  For many projects, the readme might then appear  as the 'buildInstructions', 'tutorial', 'readme' fields, while we would still miss other kinds of documentation.  At the end of the day, there simply aren't bright lines between these different kinds of documentation.  They have overlapping roles, and if we are passing the problem of disambiguating them on to the consumer, that is merely a reflection of ambiguity that was already there.  I find it difficult to articulate any use case a machine could perform by knowing only that the metadata provider had indicated a file was 'buildInstructions' rather than 'documentation'.

**Comment by dgarijo** [comment]
Thanks for clarifying that aspect about codemeta. 

I think I have three use cases where the distinction would be very helpful:
- search (UI perspective): When browsing information from multiple repositories, target the build instructions specifically. I have repos where the documentation are pdfs with > 20 pages. The build instructions are usually a separate more concise document, and it saves time to access it directly.
- learning: using the build instructions sections to train a language model to automatically extract the steps to follow in the documentation.
- machine-actionable building of the target document. However, this would only work under certain assumptions, as discussed above.

**Comment by moranegg** [comment]
FYI: https://schema.org/documentation

**Comment by moranegg** [comment]
see the new discussion to come to a consensus: https://github.com/codemeta/codemeta/discussions/335
